iroh-net/src/magic_endpoint.rs

```diff
@@ -292,12 +305,32 @@ impl MagicEndpoint {
     ///
     /// This updates the magic socket's *netmap* with these addresses, which are used as candidates
     /// when connecting to this peer (in addition to addresses obtained from a derp server).
+    ///
+    /// If no UDP addresses are added, and `derp_region` is `None`, it will error.
+    /// If no UDP addresses are added, and the given `derp_region` cannot be dialed, it will error.
     pub async fn add_known_addrs(
         &self,
         peer_id: PeerId,
+        derp_region: Option<u16>,
         endpoints: &[SocketAddr],
     ) -> anyhow::Result<()> {
-        const DEFAULT_DERP_REGION: u16 = 1;
+        match (endpoints.is_empty(), derp_region) {
+            (true, None) => {
+                anyhow::bail!(
```

Comment:
This errors if `derp_region` is `None` and `endpoints` is empty. `add_known_addrs` is currently always called by `connect`. Now if I call `connect` once with a derp region and/or a list of UDP addresses, those regions and addresses are added to the addressbook (netmap). If I then call `connect` a  second time with only a peer id, it will error even though the magicsock actually does know how to connect to the peer in question, because the addresses/regions from the first call are stored and remembered in the netmap.

This behavior is being used in #1149. When new peer info is received via a gossip `Join` or `ForwardJoin` message, this info (addresses and after this PR regions) is added to the endpoint with a direct call to `add_known_addrs`. If the gossip layer later on wants to connect to the peer, `connect` is called only with the peer id, without addresses or regions. This works, because the addresses/regions are stored in the netmap.

The nice thing about this is that we do not have to keep a separate registry of peer info in the gossip layer but instead can just use the addressbook of the magicsock.

These lines would change this behavior. Can we instead first check if the current netmap of the magicsock actually has any regions/addresses for the peerid and only error if this is empty as well?

Reply:
We could also move all these checks into the conn actor, and pass only the new info via a `Conn::add_peer_info` and `ActorMessage::AddPeerInfo` or so that would replace `ActorMessage::SetNetworkMap`. And then we could add `ActorMessage::CanDialPeer(oneshot::Sender<anyhow::Result<()>>` or similar to check if a peer is dialable and return those warnings/errors. Maybe not for this PR. But it would be nice to centralize the netmap management and not keep a copy of the netmap in the MagicEndpoint. I think the current setup might be a little brittle.

Reply:
> Can we instead first check if the current netmap of the magicsock actually has any regions/addresses for the peerid and only error if this is empty as well?

yeah, we probably should do that

Reply:
Honestly we probably should not error out here ever, but actually just try to dial with whatever we have. But this will need more refinement and testing over time

Reply:
https://github.com/n0-computer/iroh/issues/1248